### PR TITLE
Method "balance" (EnzoMethodBalance) for dynamic load balancing

### PR DESCRIPTION
**Pull request by jobordner**
This update implements a dynamic load balancer based on the Morton space-filling curve ordering available in the MethodOrderMorton method. Currently, the only useful parameter is "schedule", but more are likely to be added in the future (e.g. to support other orderings as they come online, and support other Block weightings such as particle counts.)
Preliminary tests indicate the balancer is useful, with a modest improvement of ~30% on a single-node 8-core 32^3 cosmology test run to redshift 3.0.

Documentation is in place (for the method; there are no method-specific parameters to document), and a couple of small cosmology tests, one with balancing and one without, are available in input/LoadBalance. Please see the documentation (problem_method.rst) for usage information, including a couple of potential "gotcha's".

Most of the pertinent code is in EnzoMethodBalance.?pp; other changes are mainly updating debugging code or fixing/simplifying issues related to migrating blocks.

[EDIT]
Here's a plot showing performance improvement on Frontera for weak-scaling cosmology runs (DM+gas only) showing the speedup of balanced over unbalanced. [plot-speedup.png](http://client64-71.sdsc.edu/~bordner/Pull-271/plot-speedup.png) Problem sizes are N7=128^3 N8=256^3 N9=512^3 with 4 levels of refinement (1024^3 runs are in the queue), ending at redshift 3.

[EDIT 2]
Rendered documentation: [Enzo-E Methods: "balance" method](http://client64-71.sdsc.edu/~bordner/enzo-e.jobordner.balance-cello/doc/build/html/user/problem_method.html#balance-method)

### CODE REVIEW

File: src/Cello/data_DataMsg.cpp [hunk 1]
```diff
@@ -439,50 +439,52 @@ void DataMsg::update (Data * data, bool is_local, bool is_kept)
...
             message,ifps3_cf_[0],ifps3_cf_[1],ifps3_cf_[2]);
-  CkPrintf ("%s DATA_MSG coarse ifmr3_cf_   = %d %d %d\n",
+  fprintf (fp,"%s DATA_MSG coarse ifmr3_cf_   = %d %d %d\n",
             message,ifmr3_cf_[0],ifmr3_cf_[1],ifmr3_cf_[2]);
-  CkPrintf ("%s DATA_MSG coarse ifpr3_cf_   = %d %d %d\n",
+  fprintf (fp,"%s DATA_MSG coarse ifpr3_cf_   = %d %d %d\n",
             message,ifpr3_cf_[0],ifpr3_cf_[1],ifpr3_cf_[2]);
-    
-  fflush(stdout);
+  const int ip = CkMyPe();
```
Comment (mabruzzo) [inline src/Cello/data_DataMsg.cpp:489]: Is this doing anything?

File: src/Enzo/enzo.ci [hunk 2]
```diff
@@ -199,7 +205,11 @@ module enzo {
 
     entry void p_initial_hdf5_recv(MsgInitial * msg_initial);
 
-    // EnzoMethodGravity synchronization entry methods
+    // EnzoMethodBalance
+    entry void p_method_balance_migrate();
+    entry void p_method_balance_done();
+
+  // EnzoMethodGravity synchronization entry methods
```
Comment (mabruzzo) [inline src/Enzo/enzo.ci:212]:
```suggestion
    // EnzoMethodGravity synchronization entry methods
```

File: src/Enzo/enzo_EnzoMethodBalance.cpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
+// See LICENSE_CELLO file for license and copyright information
+
+/// @file     enzo_EnzoMethodBalance.cpp
+/// @author   James Bordner (jobordner@ucsd.edu)
+/// @date     Fri Apr  2 17:05:23 PDT 2010
```
Comment (mabruzzo) [inline src/Enzo/enzo_EnzoMethodBalance.cpp:5]: Minor nitpick, the date should be updated, right?

File: src/Enzo/enzo_EnzoMethodBalance.hpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
+// See LICENSE_CELLO file for license and copyright information
+
+/// @file     enzo_EnzoMethodBalance.hpp
+/// @author   James Bordner (jobordner@ucsd.edu) 
+/// @date     Thu Apr  1 16:14:38 PDT 2010
```
Comment (mabruzzo) [inline src/Enzo/enzo_EnzoMethodBalance.hpp:5]: the date should be updated, right?

File: src/Enzo/enzo_EnzoMethodBalance.hpp [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+/// @brief    [\ref Enzo] Implementation of Enzo PPM hydro method
```
Comment (mabruzzo) [inline src/Enzo/enzo_EnzoMethodBalance.hpp:6]:
```suggestion
/// @brief    [\ref Enzo] Declaration of the the EnzoMethodBalance class
```

File: input/LoadBalance/test_balance.incl [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+         final_redshift = 3.000000000000000;
+         hubble_constant_now = 0.7000000000000000;
+         initial_redshift = 99.00000000000000;
+         max_expansion_rate = 0.01500000000000000;
+         omega_baryon_now = 0.04000000000000000;
+         omega_cdm_now = 0.2600000000000000;
+         omega_lambda_now = 0.7000000000000000;
+         omega_matter_now = 0.3000000000000000;
+     };
+     list = [ "cosmology" ];
```
Comment (mabruzzo) [inline input/LoadBalance/test_balance.incl:310]: A suggestion to use the new fluid_props physics object (I need to address this in existing parameter files)

```suggestion
     fluid_props {
         dual_energy { type = "bryan95"; }
         eos { gamma = 1.666700000000000; }
     };
     list = [ "cosmology", "fluid_props" ];
```

File: input/LoadBalance/test_balance.incl [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+         order = 2;
+         solver = "cg";
+     };
+     # Use "+=" so methods can be added to the beginning before including this file
+     list += ["ppm", "pm_deposit", "gravity", "pm_update", "comoving_expansion" ];
+     ppm {
+         courant = 0.5000000000000000;
+         diffusion = false;
+         dual_energy = true;
+     };
```
Comment (mabruzzo) [inline input/LoadBalance/test_balance.incl:143]: This is related to the suggestion about using the fluid_props physics object

```suggestion
         diffusion = false;
     };
```

File: input/LoadBalance/test_balance.incl [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+
+ Domain {
+     lower = [ 0.000000000000000, 0.000000000000000, 0.000000000000000 ];
+     upper = [ 1.000000000000000, 1.000000000000000, 1.000000000000000 ];
+ }
+
+ Field {
+     alignment = 8;
+     gamma = 1.666700000000000;
+     ghost_depth = 4;
```
Comment (mabruzzo) [inline input/LoadBalance/test_balance.incl:36]: This is related to the suggestion about using the fluid_props physics object

```suggestion
     alignment = 8;
     ghost_depth = 4;
```

**Review by mabruzzo — APPROVED** [review]
I took a quick glance and this generally looks good to me

File: src/Cello/data_DataMsg.cpp (hunk 1 above)
Comment (jobordner) [reply src/Cello/data_DataMsg.cpp:489]: residual debugging code

File: input/LoadBalance/test_balance.incl (hunk 6 above)
Comment (jobordner) [reply input/LoadBalance/test_balance.incl:310]: Deferring this to when you update other parameter files so the related updates are all in one place

**Review by WillHicks96 — APPROVED** [review]
It seems like there might be an issue with using EnzoMethodOutput for data output after load balancing steps.

See the attached test -- 'test_balance-on-bcg.txt' is the exact test in `input/LoadBalance` of the same name, but with EnzoMethodOutput for hdf5 output (I renamed it to .txt here so I could attach it). `output.txt` is the logfile from running this test, including the error message at the end. 

Note that this happens regardless of where "output" is in `Method:list`. If "output" comes before "balance", it crashes during the next step with data output.

[output.txt](https://github.com/enzo-project/enzo-e/files/10485517/output.txt)
[test_balance-on-bcg.txt](https://github.com/enzo-project/enzo-e/files/10485518/test_balance-on-bcg.txt)

Not sure if this is necessarily an issue with EnzoMethodBalance though, so hitting approve! 

Load balancing in 256^3 cosmology tests (not using EnzoMethodOutput) is working well in my tests and speeding things up quite a bit. Am also able to write checkpoint outputs using `EnzoMethodCheck` and restart from them with load balancing turned on.

**Comment by mabruzzo** [comment]
If you merge the changes from main into this branch, the failing tests should now pass

**Comment by mabruzzo** [comment]
Tests are all passing now and this has 2 approvals - I'm going to merge!

**Pull request merged**